MAINT: minimize jac docstring closes #9042

andyfaff opened the pull request:
Closes #9042

mdhaber commented:
I tested the methods:
```
import numpy as np
from scipy.optimize import minimize

def f(x):
    return np.sum(x**2)

def j(x):
    return 2*x

methods = ["COBYLA", 'Powell', 'nelder-mead',
           'L-BFGS-B', 'BFGS', 'CG', 'TNC', 'SLSQP',
           "Newton-CG", 'dogleg', 'trust-ncg', 'trust-exact',
           'trust-krylov', 'trust-constr']

jacobians = [None, j, '2-point']  # ['three-point', 'cs']
for method in methods:
    for jacobian in jacobians:
        try:
            res = minimize(f, np.random.rand(10), method=method, jac=jacobian)
            print("{0} worked with Jacobian {1}".format(method,str(jacobian)))
        except Exception as e:
            print("{0} FAILED with Jacobian {1} because {2}".format(method,str(jacobian), str(e)))
```
and got:
```
COBYLA worked with Jacobian None
COBYLA worked with Jacobian <function j at 0x0000019066F7C6A8>
COBYLA worked with Jacobian 2-point
Powell worked with Jacobian None
Powell worked with Jacobian <function j at 0x0000019066F7C6A8>
Powell worked with Jacobian 2-point
nelder-mead worked with Jacobian None
nelder-mead worked with Jacobian <function j at 0x0000019066F7C6A8>
nelder-mead worked with Jacobian 2-point
L-BFGS-B worked with Jacobian None
L-BFGS-B worked with Jacobian <function j at 0x0000019066F7C6A8>
L-BFGS-B worked with Jacobian 2-point
BFGS worked with Jacobian None
BFGS worked with Jacobian <function j at 0x0000019066F7C6A8>
BFGS worked with Jacobian 2-point
CG worked with Jacobian None
CG worked with Jacobian <function j at 0x0000019066F7C6A8>
CG worked with Jacobian 2-point
TNC worked with Jacobian None
TNC worked with Jacobian <function j at 0x0000019066F7C6A8>
TNC worked with Jacobian 2-point
SLSQP worked with Jacobian None
SLSQP worked with Jacobian <function j at 0x0000019066F7C6A8>
SLSQP worked with Jacobian 2-point
Newton-CG FAILED with Jacobian None because Jacobian is required for Newton-CG method
Newton-CG worked with Jacobian <function j at 0x0000019066F7C6A8>
Newton-CG FAILED with Jacobian 2-point because Jacobian is required for Newton-CG method
dogleg FAILED with Jacobian None because Jacobian is required for dogleg minimization
dogleg FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Hessian is required for dogleg minimization
dogleg FAILED with Jacobian 2-point because Jacobian is required for dogleg minimization
trust-ncg FAILED with Jacobian None because Jacobian is required for Newton-CG trust-region minimization
trust-ncg FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Either the Hessian or the Hessian-vector product is required for Newton-CG trust-region minimization
trust-ncg FAILED with Jacobian 2-point because Jacobian is required for Newton-CG trust-region minimization
trust-exact FAILED with Jacobian None because Jacobian is required for trust region exact minimization.
trust-exact FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Hessian matrix is required for trust region exact minimization.
trust-exact FAILED with Jacobian 2-point because Jacobian is required for trust region exact minimization.
trust-krylov FAILED with Jacobian None because ('Jacobian is required for trust region ', 'exact minimization.')
trust-krylov FAILED with Jacobian <function j at 0x0000019066F7C6A8> because Either the Hessian or the Hessian-vector product is required for Krylov trust-region minimization
trust-krylov FAILED with Jacobian 2-point because ('Jacobian is required for trust region ', 'exact minimization.')
trust-constr worked with Jacobian None
trust-constr worked with Jacobian <function j at 0x0000019066F7C6A8>
trust-constr worked with Jacobian 2-point
```
Looks like a few other methods require a Jacobian: `trust-krylov`, `trust-exact`, and `dogleg`.

mdhaber commented:
There are several formatting issues in the `minimize` documentation.
https://14717-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.optimize.minimize.html#scipy.optimize.minimize

![image](https://user-images.githubusercontent.com/6570539/63186446-48363580-c011-11e9-8938-1f2edb508275.png) 
(shouldn't have vertical gray bar such places - caused by incorrect indentation)

![image](https://user-images.githubusercontent.com/6570539/63186503-6dc33f00-c011-11e9-8e7d-e335a6899c45.png)
`x`, `args`, `jac` need double back-ticks to render as teletype in RST.



Do you want to fix those while you're in there, or save for separate PR?

mdhaber commented:
Let me know if I should open a separate issue for all this.

mdhaber commented on scipy/optimize/_minimize.py:
```diff
@@ -91,10 +91,11 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
-        {'2-point', '3-point', 'cs'} select a finite
-        difference scheme for numerical estimation of the gradient. Options
-        '3-point' and 'cs' are available only to 'trust-constr'.
+        the fixed parameters. Newton-CG and trust-ncg require that a callable
```
```suggestion
        the fixed parameters. Methods 'Newton-CG', 'dogleg', 'trust-ncg', 'trust-exact', and 
        'trust-krylov'  require that a callable
```

andyfaff replied:
done

andyfaff commented:
@mdhaber I've updated `jac` to refer to the extra methods requiring a callable jacobian. I also added the backticks you requested.

> Looks like a few other methods require a Jacobian: trust-krylov, trust-exact, and dogleg. These should probably be noted here, too.

done

> x, args, jac need double back-ticks to render as teletype in RST.

done

> "L-BFGS-B worked with Jacobian 3-point" (several methods work with '3-point', 'cs', but are they actually using those methods to get Jacobian information?)

If you trawl through the code you'll see that they work, but they should [issue a warning](https://github.com/scipy/scipy/blob/master/scipy/optimize/_minimize.py#L547) that they're just going to use 2-point forward differences anyway.

andyfaff commented:
Don't merge this just yet, I'm not sure if there is a mistake in the `minimize` function.

andyfaff commented:
Closed, including in omnibus #10673
